dashboard/src/actions/authActions.js

```diff
@@ -7,6 +7,161 @@ import Cookies from "js-cookie";
 import { SUCCESS } from "assets/constants/overviewConstants";
 import { showToast } from "actions/toastActions";
 import { uid } from "../utils/helper";
+import { fetchEndpoints } from "actions/endpointAction";
+
+
+/**
+ * Get the key value.
+ * @param {key} key can be a string.
+ * @return {value} a string value of a key
+ */
+export function getQueryVariable (key) {
+    const query = window.location.search.substring(1);
+    const vars = query.split('&');
+    let code = "";
+    for (let i = 0; i < vars.length; i++) {
+      const pair = vars[i].split('=');
+      if (pair[0] === key) {
+        code = pair[1];
+      }
+      // else return "no-idea";
+    }
+    return code;
+
+}
+
+/**
+ * Get user details from the OIDC server.
+ * @param {endpoints} endpoints from the getState.
+ * @return {user} a dict value of a user details obtained from the token
+ */
+export function getOIDCUserDetails(endpoints) {
+    const reqUser = new XMLHttpRequest();
+    const OIDCToken = Cookies.get("token");
+    const oidcServer = endpoints.openid.server;
+    const oidcRealm = endpoints.openid.realm;
+    const uri = `${oidcServer}/realms/${oidcRealm}/protocol/openid-connect/userinfo`;
+    let userPayload = {};
+    return new Promise((resolve, reject) => {
+      reqUser.onreadystatechange = function() {
+        if (reqUser.readyState === 4) {
+          const response = JSON.parse(reqUser.responseText);
+          userPayload = {
+            "username": response?.preferred_username,
+            "email": response?.email,
+            "first_name": response?.given_name,
+            "last_name": response?.family_name,
+          };
+          resolve(userPayload);
+        }
+      }
+      reqUser.open('GET', uri, false);
+      reqUser.setRequestHeader('Content-type', 'application/json');
+      reqUser.setRequestHeader('Authorization', 'Bearer ' + OIDCToken);
+      reqUser.send();
+     });
+}
+
+export const loadTokens = () => async (dispatch, getState) => {
+    const code = getQueryVariable('code');
+
+    if (code !== "") {
+      let endpoints = getState().apiEndpoint.endpoints;
+      if (!("openid" in endpoints)){
+          await dispatch(fetchEndpoints);
+      }
+      endpoints = getState().apiEndpoint.endpoints;
```

Comment:
> that doesn't seem like an operational mode we would need to support

No...let's defer that mindset until someone requires it.

For now, let's assume that every time the Dashboard hits the server for the endpoints, the answer will be the same, so it doesn't need to reload them during a session (and let's ignore the transition which might occur when the Server is restarted or upgraded).

Reply:
Yeah, the problem is after getting redirected back (to auth page) from the Keycloak broker, the redux seems to lose the current state (which we refer to using `getState()`). After redirecting back, the `endpoint` state is an empty dictionary. So I had to fetch the endpoints again and I have to use the `await` because unless all the endpoints are loaded I can not refer to the `endpoints.openid` parameter. 

Another alternative I tried was to use `useEffect` functionality of the redux to fetch endpoints on the `AuthForm` component which makes an asynchronous call to fetch endpoints. But because these are asynchronous there is no guarantee that the endpoints will be loaded before we refer to the  `endpoints.openid` parameter to make the OIDC token request (and as expected most of the time this fails).
 
I am sure there is a better way to store the current state so they are not lost upon redirection but I am not aware of that at the moment.

Reply:
That's a good point. Yeah, the redirect back to the dashboard is actually a *reload* of the React application from the server, so I guess we shouldn't expect the browser to retain any of the local context: and maybe it's even best that it doesn't. Huh.

Reply:
> the redirect back to the dashboard is actually a reload of the React application from the server, so I guess we shouldn't expect the browser to retain any of the local context

Hmmm...I wonder what _else_ that's breaking for us.  🙄 

So, if we're reloading the application from scratch, why doesn't its initialization load the endpoints like it does on the initial load??

(I would really like to understand this situation before we start trying to hack our way around it.)

Reply:
> So, if we're reloading the application from scratch, why doesn't its initialization load the endpoints like it does on the initial load??

It does eventually, so that's why I have a conditional statement here that checks for `"openid" in endpoints` (meaning whether the endpoints are already loaded). However, If you look under `App.js` the `fetchEndpoints` call is made under the redux asynchronous artifact `useEffect()`, and then it returns all the components like `LoginForm`, `AuthForm`, etc. 

Now, if we are already on the Auth page and the asynchronous call to `fetchEndponts` haven't been completed yet then I have to wait for `fetchEndpoint` to complete because the OIDC `loadToken` call is dependent on `endpoints.openid` value. So yes there are 2 Pbench endpoints API requests happening, one here and one from the asynchronous App.js. 

I tried using the arbitrary `wait` but that wasn't a very clean solution and I tried using `while true` until the endpoints are loaded but that seems to hang the dashboard which I still need to understand why. In the meantime, this felt like the most straightforward solution.

Reply:
> I tried using the arbitrary `wait` but that wasn't a very clean solution and I tried using `while true` until the endpoints are loaded but that seems to hang the dashboard which I still need to understand why. In the meantime, this felt like the most straightforward solution.

`while true` would be very unfriendly.  (See [here](https://developer.mozilla.org/en-US/docs/Web/API/HTML_DOM_API/Microtask_guide/In_depth#problems), and in the surrounding page.)  If you want to do something like this, then I believe you need to do it in a subroutine which you can schedule to run again (e.g., after a timeout of "0" or perhaps longer) if things aren't ready.

> I tried using the arbitrary `wait`

I'm not sure what you mean by that.  Was that something like `await null` or some sort of Javascript equivalent of `sleep()`? 

Ideally, I think we would come up with a way for your code to access the `Promise` returned by the original call to `dispatch(fetchEndpoints)` in `App()`, but the details of that are unpalatable.  So, I think I would try something like this:
```
async function wait_for_endpoints() {
  const wait_start = Date.now();
  const max_wait = 10000;  // Milliseconds

  function check(resolve, reject) {
    if (Object.keys(getState().apiEndpoint.endpoints).length !== 0) {
      resolve("Loaded");
    } else if (Date.now() - wait_start > max_wait) {
      reject("Wait for endpoint exceeded ${max_wait} milliseconds");
    } else {
      setTimeout(check, 250, resolve, reject);
    }
  };

  return new Promise(check)  
}
```
The value specified to `setTimeout()` doesn't matter, as long as it's not too big (e.g., zero would be a perfectly acceptable value) -- `check()` won't run until the next pass of the event-loop, so small values won't block anything else from running.

So, instead of conditionally `await`'ing a `dispatch()` for `fetchendpoints`, you would use something like this:
```
      await wait_for_endpoints();
      endpoints = getState().apiEndpoint.endpoints;
```
which avoids the redundant fetch and state updates.

Reply:
@webbnh I tried the one you suggested but what's happening is I think this simply wrapping the promise. The `await` inside the async function does not wait for anything when you run it, it just returns a promise. I will extend from your suggestion and see what I can do to make it wait for it to `resolve`.

Reply:
Per [the docs](https://developer.mozilla.org/en-US/docs/Learn/JavaScript/Asynchronous/Promises#async_and_await), 

> Inside an async function, you can use the `await` keyword before a call to a function that returns a promise. This makes the code wait at that point until the promise is settled, at which point the fulfilled value of the promise is treated as a return value, or the rejected value is thrown.

So, if you're seeing it not stop at the `await wait_for_endpoints()` line, I bet that there is an exception being thrown.  When I was experimenting with the above code, I found that exceptions don't propagate (i.e., become "visible") beyond the `await` statement.  Instead, you need to use a `.catch()` on the promise which `wait_for_endpoints()` returns.  (I wonder if the wait is hitting the max timeout quicker than you think.)